Remove uses_master_app_form_ids

orangejenny opened the pull request:
##### SUMMARY
Small change, long backstory. This is # 4 in https://github.com/dimagi/commcare-hq/issues/25718

`xforms_index` is a couch view that stores the unique ids of all forms. Its only purpose is to look up forms by id; it's the enforcer of form id global uniqueness.

A year ago, a [report builder bug](https://manage.dimagi.com/default.asp?283130) was reported where form-based reports could not be created based on linked app forms because linked apps were not in `xforms_index`. https://github.com/dimagi/commcare-hq/pull/21872 fixed this by adding linked apps to `xforms_index`.

A few days later, a [form builder bug](https://manage.dimagi.com/default.asp?283410) was reported where the app properties window wouldn't load, because the ids in `xforms_index` were no longer unique, and app properties depended on `xforms_index` to look up the form by id. https://github.com/dimagi/commcare-hq/pull/21969/ fixed this by adding a flag, `uses_master_app_form_ids`, to the linked application model and updating `xforms_index` to exclude apps with that flag set. This would re-introduce the report builder bug for any apps with this flag on. The PR included a migration to turn this flag on for all applications except the domain that had reported the report builder bug, with the perhaps optimistic [plan](https://github.com/dimagi/commcare-hq/pull/21969/files#r221619379) that solutions would examine these domains individually and turn the flag off.

This week, https://github.com/dimagi/commcare-hq/pull/25725 was done to remove report builder's dependency on `xforms_index`, so the original report builder bug is no longer an issue, and https://github.com/dimagi/commcare-hq/pull/25724 was done to remove form builder's dependency on `xforms_index`, so the app properties bug is also no longer an issue.

This PR removes the `uses_master_app_form_ids` flag. It leaves linked applications in `xforms_index`, because in https://github.com/dimagi/commcare-hq/pull/24945 I added support for SMS surveys to linked apps, and SMS surveys do still depend on `xforms_index`, although that is planned for removal as part of the [proposal to make form ids no longer globally unique](https://github.com/dimagi/commcare-hq/issues/25718).

orangejenny committed:
Removed uses_master_app_form_ids from model, couch view, and existing linked apps

emord reviewed (commented):
This seems reasonable though scary and from the backstory it certainly looks like you've done the necessary ground work. Couple of questions about this though:

1. Does the migration also need to go through builds or LinkedApplication builds (if those are different?)?
2. Is there any concern that we don't have a way to revert this? Should the migration at least write to a file to say what apps were changed?
3. When these linked apps are added to the xforms_index, what's stopping the SMS surveys from erroring  at https://github.com/dimagi/commcare-hq/blob/f347e6a3362df2c3daa936282fdf16f6c01c5d24/corehq/apps/app_manager/models.py#L1087 ?

snopoke reviewed (commented):
Surely this can't be changed until all usages of the view that require uniqueness are removed?

orangejenny commented:
@emord @snopoke Thanks for feedback - think I got overexcited about removing this. I'll revisit this when then various workflows that require form id uniqueness are updated. Marking don't merge for now.

orangejenny committed:
Merge branch 'master' into jls/kill-uses_master_app_form_ids

orangejenny commented:
The workflows that require uniqueness are all now fixed, and `xforms_index` itself has been removed in https://github.com/dimagi/commcare-hq/pull/25776 . I think this is now good to go, it's just the migration...it really doesn't do much, I just spent enough time investigating this that I'd like to fully clean it up.

@emord 

> Does the migration also need to go through builds or LinkedApplication builds (if those are different?)?

It goes through all `LinkedApplication` and `LinkedApplication-Deleted` docs, so both apps and builds are covered.

> Is there any concern that we don't have a way to revert this? Should the migration at least write to a file to say what apps were changed?

I don't think this flag has any real meaning that couldn't be acquired another way. It's a bunch of apps that at one point in time may have been subject to the report builder bug, which is now fixed.

> When these linked apps are added to the xforms_index, what's stopping the SMS surveys from erroring at...

This is no longer an issue with `xforms_index` gone.

emord commented on corehq/apps/app_manager/migrations/0008_remove_uses_master_app_form_ids.py:
```diff
@@ -0,0 +1,40 @@
+def _pop_deprecated_field(apps, schema_editor):
+    app_ids = (get_doc_ids_by_class(LinkedApplication)
+               + get_deleted_doc_ids_by_class(LinkedApplication))
+    iter_update(LinkedApplication.get_db(), _pop_field, with_progress_bar(app_ids), chunksize=1)
```
this should probably be fast (due to few docs) on all environments except prod, india and icds which may have >10k docs (haven't checked, but just thinking through how many there will be for icds). I think that might end up being pretty long to block deploy for. Are you planning to run this outside of a normal deploy for these environments?

orangejenny replied:
I had not checked, just did. There aren't that many docs: 2150 on prod, 150 on india, 300 on icds. India doesn't have many linked apps, and on icds it looks like if you go back far enough AWW builds aren't linked apps, they're regular `Application` docs. Does that seem low enough to do in a regular deploy?

I suppose the correct answer is if I don't know how long it'll take I should run this outside of deploy.

emord replied:
Oh interesting I would have definitely expected more on india and icds. Seems worth running on prod and then making a call on whether or not to do it on the others

orangejenny replied:
sounds good

emord reviewed (commented):
Thank you for completing the migration and ensuring a lazy migration doesn't stick around for long.

orangejenny commented:
Ran the migration in a private release on prod, india, and icds. Didn't take long. Merging.
